### ISSUE TITLE
Cannot toggle series visibility when loading `highcharts-more.js` dynamically

### ISSUE THREAD

**Issue by nishikantparmariam** [highcharts/highcharts#22168]
#### Expected behaviour
Series visibility toggling on a line series via legend should work as expected.

#### Actual behaviour

As explained in https://github.com/highcharts/highcharts/issues/22065, we load Highcharts module on the fly depending upon the series type. 

It seems that loading `highcharts-more.js` (for boxplot series) later after loading `highcharts.js` (for line series) creates troubles in toggling the visibility of the line series.

In below reproducer, the visibility of line series (in the first chart) is off initially and then when we try to toggle it via the legend we get below error in the console:

![image](https://github.com/user-attachments/assets/eaa440e8-a186-4a7f-8f88-4b0198f1df16)

#### Live demo with steps to reproduce

**https://jsfiddle.net/vgatxqLj/**

Click on legend in first chart and observe that the series is still hidden and notice the error in console.

#### Product version
Highcharts 11.4.8

#### Affected browser(s)
Tried on Chrome

**Comment by karolkolodziej** [comment]
Hi @nishikantparmariam!
Thank you for pointing that out.

_Internal note:_
The `afterBuildStacks` gets triggered for the chart that doesn't have the waterfall axis loaded.

_Workaround:_
You can check if the waterfall axis exists by removing the event and adding the correct one that checks if the `waterfall` exists- [demo](https://jsfiddle.net/BlackLabel/2w6qrbaL/).
```TS
  function onAxisAfterBuildStacks() {
    const axis = this,
      stacks = axis.waterfall?.stacks

    if (stacks) {
      stacks.changed = false
      delete stacks.alreadyChanged
    }
  }
  Highcharts.removeEvent(Highcharts.Axis, "afterBuildStacks")
  Highcharts.addEvent(
    Highcharts.Axis,
    "afterBuildStacks",
    onAxisAfterBuildStacks,
  )
```

**Comment by nishikantparmariam** [comment]
Thanks @karolkolodziej, this worked!

**Comment by Xillez** [comment]
_Internal note:_
The main issue is that initialization events (e.g. 'init' in this case, but also possibly others in other modules) happen before the module begins to listen for them, and never again which leads to undefined properties. So a full and proper fix might be some kind of post-initialization step each module can implement and run to get already existing charts up to speed, if needed.

I can make a PoC for this specific case but I think a better and more proper solution needs to be implemented later. This fix might also be made deprecated if Torstein finishes the dynamic loading feature mentioned in the linked issue above (#14478). Any thoughts @TorsteinHonsi?

**Comment by TorsteinHonsi** [comment]
> Any thoughts @ Torstein Honsi?

Yes, without digging into it: I'm a bit surprised that events happen before the module starts listening to them, because series initialization runs when all modules are loaded. Is this true events added with `addEvent`, or is it those event-like `wrap` statements? If they are wraps, we could try rewriting it to use `addEvent`. Where is the code in question, `init` in this case?

> better and more proper solution needs to be implemented later. This fix might also be made deprecated if Torstein finishes the dynamic loading feature

Maybe, but the autoloader also just asynchronously loads script, much like the OP demo. So we may have the same problem there.

**Comment by Xillez** [comment]
This code =>
https://github.com/highcharts/highcharts/blob/abbf5a7153b2c3e24cecfd9e8f45c9aa5a416608/ts/Core/Axis/WaterfallAxis.ts#L98-L111

In the jsfiddel linked in the original issue comment above, the first chart initializes, then the script element in made and injected, script gets loaded asynchronously and once fully loaded the next chart is made. **But**, the first chart misses out on the init event added in the WaterfallAxis's compose function but all the other events fire like normal, leading to undefined properties.

How I found out:
Adding a quick console log to WaterfallAxis's onAxisInit like below makes it only print for the second chart, not the first. 
```
function onAxisInit(this: Axis): void {
    const axis = this;
    if (!axis.waterfall) {
        axis.waterfall = new Composition(axis as WaterfallAxis);
        console.log(`CREATED axis.waterfall for chart with index: ${this.chart.index}`, axis);
   }
}
```

But if i run the init function for all preexisting charts the issue is fixed (don't know if there might be some kind of lingering issues by doing this with other modules).
```
    export function compose(
        AxisClass: typeof Axis,
        ChartClass: typeof Chart,
        charts?: Array<Chart>
    ): void {

        if (pushUnique(composed, 'Axis.Waterfall')) {

            // #22168, trigger init "event" for all existing axes if needed.
            if (charts && charts.length) {
                charts.forEach(function (chart: Chart): void {
                    chart.axes.forEach(function (axis: Axis): void {
                        if (!axis.waterfall) {
                            onAxisInit.call(axis);
                        }
                    });
                });
            }

            addEvent(AxisClass, 'init', onAxisInit);
            addEvent(AxisClass, 'afterBuildStacks', onAxisAfterBuildStacks);
            addEvent(AxisClass, 'afterRender', onAxisAfterRender);

            addEvent(ChartClass, 'beforeRedraw', onChartBeforeRedraw);
        }
    }
```

Without fix:
![WithoutFix](https://github.com/user-attachments/assets/59680798-1db4-4769-9395-5d8251640561)

With fix:
![WithFix](https://github.com/user-attachments/assets/cd7368c7-3977-4b9a-ae4d-f2f67c8551f1)

Please keep asking if things aren't fully clear still 👍

**Comment by aditya3434** [comment]
I've encountered a somewhat similar issue with column chart. Toggling chart visibility exhibits different behavior if `highcharts-more.js` is imported dynamically.

#### jsFiddle example
https://jsfiddle.net/a6tqu3zj/

#### Bug Showcase
![Image](https://github.com/user-attachments/assets/46c9a8e5-3a29-44c9-a5b1-33fbe99b50b7)

**Comment by TorsteinHonsi** [comment]
Simplified demo, local code: https://jsfiddle.net/highcharts/tpuke829/

**Comment by TorsteinHonsi** [comment]
@aditya3434 Sorry for the late reply! I created a fix in #23057.

### Demos with the fix applied
* The original post: https://jsfiddle.net/highcharts/8atqypvL/
* The posted demo from two weeks ago: https://jsfiddle.net/highcharts/3e7859tp/

**Comment by TorsteinHonsi** [comment]
@aditya3434 Yes, it actually turned out pretty simple. We just add a mock `waterfall` object to the Axis prototype.

```js
// Workaround for #22168, issue after dynamically loading highcharts-more.js
Highcharts.Axis.prototype.waterfall = {
	stacks: {}
};
```

* Augmented OP demo: https://jsfiddle.net/highcharts/4ftea0x2/
* Augmented error bar demo: https://jsfiddle.net/highcharts/0mgcnqfo/
